### ISSUE TITLE
not in filter confuses guard against None values

### ISSUE THREAD

**Issue by wakemaster39** [python/mypy#6305]
Note: if you are reporting a wrong signature of a function or a class in
the standard library, then the typeshed tracker is better suited
for this report: https://github.com/python/typeshed/issues

Please provide more information to help us understand the issue:

* Are you reporting a bug, or opening a feature request?

Bug

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```
import re
from typing import Optional, Any


def test(y: Optional[str]) -> Any:
    if y:
        return lambda x: not re.search(y,x) # This is fine

def test2(y: Optional[str]) -> Any:
    if y:
        return filter(lambda x: not re.search(y,x), ["1", "2"]) #raises an error
```

* What is the actual behavior/output?
test.py:11:37: error: Value of type variable "AnyStr" of "search" cannot be "Optional[str]"

* What is the behavior/output you expect?
There should be no errors

* What are the versions of mypy and Python you are using?
  Do you see the same issue after installing mypy from Git master?
running 660, issue exists in master

* What are the mypy flags you are using? (For example --strict-optional)
[mypy]
```
# Specify the target platform details in config, so your developers are
# free to run mypy on Windows, Linux, or macOS and get consistent
# results.
python_version=3.6
platform=linux

# flake8-mypy expects the two following for sensible formatting
show_column_numbers=True

# show error messages from unrelated files
follow_imports=normal

# suppress errors about unsatisfied imports
ignore_missing_imports=True

# be strict
disallow_untyped_calls=True
warn_return_any=True
strict_optional=True
warn_no_return=True
warn_redundant_casts=True
warn_unused_ignores=True

# The following are off by default.  Flip them on if you feel
# adventurous.
disallow_untyped_defs=True
check_untyped_defs=True

# No incremental mode
cache_dir=/dev/null
```

**Comment by lincolnq** [comment]
I reduced your test case a bit and added some calls to reveal_type, which give a clue about what is going on. Here's my test case:

```
from typing import Optional, Any, List, Callable

def test2(ostr: Optional[str]) -> Any:
    if ostr:
        z: List[str]
        lambda x: reveal_type(ostr)
        filter(lambda x: reveal_type(ostr), z)
```

And here's the output:
```
foo.py:6: error: Revealed type is 'builtins.str'
foo.py:7: error: Revealed type is 'builtins.str'
foo.py:7: error: Revealed type is 'Union[builtins.str, None]'
```

Surprisingly, there are only 2 calls to reveal_type, and yet 3 `Revealed type is` appear in the output. I don't really fully understand type narrowing very well but it seems clear it is involved: if you remove the "if ostr" line, the duplicated reveal_type disappears.

**Comment by ilevkivskyi** [comment]
> Surprisingly, there are only 2 calls to reveal_type

I think this is because mypy infers types for lambdas as arguments in two passes. Also maybe the cause of the bug is that the type restriction is "lost" between the two passes? This needs some more investigation.

**Comment by eirnym** [comment]
I'd add one more example where for some unknown for me reason, `mypy` prints an error:

```
error: Argument 1 to "filter" has incompatible type "Callable[[Any], bool]"; expected "Callable[[tuple[str, str | None]], TypeGuard[tuple[str, str]]]"  [arg-type]
```


```python
def filter_none(
    values: abc.Sequence[tuple[str, str]]
) -> abc.Iterable[tuple[str, str | None]]:
    return filter(lambda value: value[1] is not None, values) # error is here
```

**Comment by ilevkivskyi** [comment]
This now works correctly.
